grpc/dispatcher.go

```diff
@@ -0,0 +1,180 @@
+package grpc
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/gojek/fiber"
+	fiberError "github.com/gojek/fiber/errors"
+	"github.com/gojek/fiber/protocol"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/credentials/insecure"
+	"google.golang.org/grpc/metadata"
+	"google.golang.org/grpc/reflection/grpc_reflection_v1alpha"
+	"google.golang.org/grpc/status"
+	"google.golang.org/protobuf/proto"
+	"google.golang.org/protobuf/reflect/protodesc"
+	"google.golang.org/protobuf/reflect/protoreflect"
+	"google.golang.org/protobuf/reflect/protoregistry"
+	"google.golang.org/protobuf/types/descriptorpb"
+	"google.golang.org/protobuf/types/dynamicpb"
+)
+
+const (
+	TimeoutDefault = time.Second
+)
+
+type Dispatcher struct {
+	timeout time.Duration
+	// serviceMethod is the service and method of server point in the format "{grpc_service_name}/{method_name}"
+	serviceMethod string
+	// endpoint is the host+port of the grpc server, eg "127.0.0.1:50050"
+	endpoint string
+	// conn is the grpc connection dialed upon creation of dispatcher
+	conn *grpc.ClientConn
+	// ResponseProto is the proto return type of the service.
+	responseProto proto.Message
+}
+
+type DispatcherConfig struct {
+	Service  string
+	Method   string
+	Endpoint string
+	Timeout  time.Duration
+}
+
+func (d *Dispatcher) Do(request fiber.Request) fiber.Response {
+	grpcRequest, ok := request.(*Request)
+	if !ok {
+		return fiber.NewErrorResponse(
+			fiberError.FiberError{
+				Code:    int(codes.InvalidArgument),
+				Message: "fiber: grpc dispatcher: only grpc.Request type of requests are supported",
+			})
+	}
+
+	ctx, cancel := context.WithTimeout(context.Background(), d.timeout)
+	defer cancel()
+	ctx = metadata.NewOutgoingContext(ctx, grpcRequest.Metadata)
+
+	responseProto := proto.Clone(d.responseProto)
+	var responseHeader metadata.MD
+	err := d.conn.Invoke(ctx, d.serviceMethod, grpcRequest.Payload(), responseProto, grpc.Header(&responseHeader))
+	if err != nil {
+		// if ok is false, unknown codes.Unknown and Status msg is returned in Status
+		responseStatus, _ := status.FromError(err)
+		return fiber.NewErrorResponse(
+			fiberError.FiberError{
+				Code:    int(responseStatus.Code()),
+				Message: responseStatus.String(),
+			})
+	}
+
+	return &Response{
+		Metadata: responseHeader,
+		Message:  responseProto,
+		Status:   *status.New(codes.OK, "Success"),
+	}
+}
+
+// NewDispatcher is the constructor to create a dispatcher. It will create the clientconn and set defaults.
+// Endpoint, serviceMethod and response proto are required minimally to work.
+func NewDispatcher(config DispatcherConfig) (*Dispatcher, error) {
```

Comment:
can you refactor this function to make it easier to read? I think extracting some of the logics into its own internal function will help to reduce the function's complexity.